File: agent/lib/tick-strategy.js
// agent/lib/tick-strategy.js — P4: the REFERENCE ORACLE for
// tick_momentum_breakout v1 (docs/tick-momentum/plan.md §5, §6; register
// TM-01, TM-03, TM-04). A deliberately plain implementation — every window
// recomputed from arrays on every event — whose only job is to be easy to
// audit and to agree, signal for signal, with the fast incremental C++
// implementation (cpp-exec/src/tick_strategy.cpp) on the checked-in
// fixtures. It depends on nothing but quotes: no candles, no LLM, no clock
// beyond the receive time used for the staleness bound.
//
// Units. Prices are wire integers (cTrader's 1e-5). The midpoint is kept
// as TWICE-mid (bid + ask) so half increments stay integers; ranges,
// movement D, volatility V and the buffer B are in twice-mid units; the
// stop distance is returned in price units (divided by two).
//
// The accepted event: a two-sided, uncrossed, non-snapshot quote whose bid
// or ask changed, with spread within the absolute ceiling and receive gap
// within the age bound. Everything else warms nothing and counts nowhere.

import { createHash } from 'node:crypto'

export const STRATEGY_ID = 'tick_momentum_breakout'
export const STRATEGY_VERSION = 'v1'

export const DEFAULT_PARAMS = Object.freeze({
  rangeEvents: 256,            // N
  momentumEvents: 64,          // M (N/4)
  minEfficiency: 0.4,          // E
  spreadBufferMult: 0.5,       // k_spread
  confirmations: 2,            // events including the crossing
  stopVolMult: 2,              // stop = max(stopVolMult × V / 2, minStop) in price units
  minStopPrice: 1,             // broker minimum distance (wire units)
  priceIncrement: 1,           // permitted price increment (wire units) — from broker metadata, never the wire scale
  maxSpread: 1000000,          // absolute spread ceiling (wire units)
  maxQuoteAgeMs: 60000,        // receive gap that invalidates the setup
  expiryEvents: 256,           // N: untriggered setup expires
  rearmCooldownEvents: 64,     // N/4
})

export function normalizeParams(p = {}) {
  const out = { ...DEFAULT_PARAMS, ...p }
  if (p.rangeEvents && p.momentumEvents == null) out.momentumEvents = Math.max(1, Math.floor(out.rangeEvents / 4))
  if (p.rangeEvents && p.expiryEvents == null) out.expiryEvents = out.rangeEvents
  if (p.rangeEvents && p.rearmCooldownEvents == null) out.rearmCooldownEvents = Math.max(1, Math.floor(out.rangeEvents / 4))
  return out
}

/** The profile hash: strategy id + version + the parameters, canonical order. */
export function profileHash(params) {
  return profileHashFull(params).slice(0, 16)
}

/** P5: the full sha256 (64 hex) the engine record pins (ENGINE_STATUS_SHAPE
 *  profileHash is HEX64); the sidecar and the trial ledger print its first
 *  16 characters, so a pinned record matches a reported profile by prefix. */
export function profileHashFull(params) {
  const p = normalizeParams(params)
  const canon = JSON.stringify({ id: STRATEGY_ID, version: STRATEGY_VERSION, ...Object.fromEntries(Object.keys(p).sort().map(k => [k, p[k]])) })
  return createHash('sha256').update(canon).digest('hex')
}

/** The profile id the engine record carries beside the hash: strategy@version. */
export const PROFILE_ID = `${STRATEGY_ID}@${STRATEGY_VERSION}`

function median(arr) {
  const s = [...arr].sort((a, b) => a - b)
  const n = s.length
  return n === 0 ? 0 : (n % 2 ? s[(n - 1) / 2] : Math.floor((s[n / 2 - 1] + s[n / 2]) / 2))
}

/**
 * One symbol's strategy. feed(q) with q = { seq, recvMs, bid, ask, snapshot,
 * crossed, changed } (bid/ask null when absent) → a signal object or null.
 * `state` is readable for tests: WARMING | ARMED | CONFIRMING | SIGNALLED | EXPIRED.
 */
export class TickMomentumOracle {
  constructor(params = {}) {
    this.p = normalizeParams(params)
    this.hash = profileHash(this.p)
    this.mids = []     // accepted twice-mids, oldest first
    this.bids = []
    this.asks = []
    this.spreads = []
    this.accepted = 0
    this.lastRecvMs = null
    this.state = 'WARMING'
    this.setup = null   // frozen at arm: { H, L, bidHigh, bidLow, askHigh, askLow, B, armedAt, id, confirmed, dir }
    this.setupSeq = 0
    this.signalledAt = null
    this.expiredAt = null
    this.rejected = { invalid: 0, spread: 0, stale: 0, repeat: 0 }
  }

  _accept(q) {
    if (q.bid == null || q.ask == null || q.snapshot || q.crossed) { this.rejected.invalid++; return false }
    if (!q.changed) { this.rejected.repeat++; return false }
    if (q.ask - q.bid > this.p.maxSpread) { this.rejected.spread++; return false }
    if (this.lastRecvMs != null && q.recvMs - this.lastRecvMs > this.p.maxQuoteAgeMs) { this.rejected.stale++; return 'stale' }
    return true
  }

  _invalidate() {
    // Invalid or stale data invalidates the setup and the warm-up.
    this.mids = []; this.bids = []; this.asks = []; this.spreads = []
    this.setup = null; this.state = 'WARMING'
  }

  feed(q) {
    const { p } = this
    const acc = this._accept(q)
    if (acc === 'stale') { this._invalidate(); this.lastRecvMs = q.recvMs; return null }
    if (!acc) {
      // Plan §4/§5: a snapshot (or a continuity break marked as one), a
      // one-sided update or a crossed quote invalidates the setup and the
      // warm-up; a repeat or an over-wide spread merely does not count.
      if (q.bid == null || q.ask == null || q.snapshot || q.crossed) this._invalidate()
      return null
    }
    this.lastRecvMs = q.recvMs
    const N = p.rangeEvents, M = p.momentumEvents
    const mid2 = q.bid + q.ask
    const spread = q.ask - q.bid
    // Windows over the PRIOR events (the candidate is excluded).
    const priorMids = this.mids
    const warm = priorMids.length >= N + 1 && priorMids.length >= M
    let signal = null
    if (warm) {
      const lastN = priorMids.slice(-N)
      const H = Math.max(...lastN), L = Math.min(...lastN)
      const diffsN = []
      for (let i = priorMids.length - N; i < priorMids.length; i++) diffsN.push(priorMids[i] - priorMids[i - 1])
      const V = Math.sqrt(M * diffsN.reduce((s, d) => s + d * d, 0) / diffsN.length)
      const medSpread = median(this.spreads.slice(-N))
      const B = Math.max(4 * p.priceIncrement, Math.round(p.spreadBufferMult * 2 * medSpread))
      // Movement over M events uses the candidate's own mid and diff.
      const seq = [...priorMids.slice(-M), mid2]
      const D = mid2 - seq[0]
      let sumAbs = 0
      for (let i = 1; i < seq.length; i++) sumAbs += Math.abs(seq[i] - seq[i - 1])
      const E = sumAbs > 0 ? Math.abs(D) / sumAbs : 0
      this.accepted++
      const ev = this.accepted

      if (this.state === 'SIGNALLED') {
        const s = this.setup
        if (mid2 >= s.L && mid2 <= s.H && ev - this.signalledAt >= p.rearmCooldownEvents) { this.setup = null; this.state = 'WARMING' }
      } else if (this.state === 'EXPIRED') {
        if (ev - this.expiredAt >= Math.max(1, Math.floor(p.rangeEvents / 4))) { this.setup = null; this.state = 'WARMING' }
      }

      if (this.state === 'WARMING' && this.setup == null) {
        // Arm only inside a positive prior range.
        if (H > L && mid2 >= L && mid2 <= H && Number.isFinite(V) && V > 0) {
          const lastBids = this.bids.slice(-N), lastAsks = this.asks.slice(-N)
          this.setup = { H, L, bidHigh: Math.max(...lastBids), bidLow: Math.min(...lastBids), askHigh: Math.max(...lastAsks), askLow: Math.min(...lastAsks), B, armedAt: ev, id: ++this.setupSeq, confirmed: 0, dir: null }
          this.state = 'ARMED'
        }
      } else if (this.state === 'ARMED' || this.state === 'CONFIRMING') {
        const s = this.setup
        if (ev - s.armedAt > p.expiryEvents && this.state === 'ARMED') {
          this.state = 'EXPIRED'; this.expiredAt = ev
        } else {
          const longOk = mid2 > s.H + s.B && D > 0 && E >= p.minEfficiency && q.bid > s.bidHigh
          const shortOk = mid2 < s.L - s.B && D < 0 && E >= p.minEfficiency && q.ask < s.askLow
          const dir = longOk ? 'BUY' : shortOk ? 'SELL' : null
          if (dir && (s.dir == null || s.dir === dir)) {
            s.dir = dir
            s.confirmed += 1 // the crossing counts as confirmation event 1
            this.state = 'CONFIRMING'
            if (s.confirmed >= p.confirmations) {
              const stopDist = Math.max(p.minStopPrice, Math.round(p.stopVolMult * V / 2))
              // PR-D: the direction is stated where it is decided — the frozen boundary the mid broke.
              signal = { symbolId: q.symbolId ?? null, side: dir, dirReason: dir === 'BUY' ? 'tick:break_high' : 'tick:break_low', seq: q.seq, recvMs: q.recvMs, trigger2: mid2, bid: q.bid, ask: q.ask, stopDistance: stopDist, spread, V: +V.toFixed(6), D, E: +E.toFixed(6), H: s.H, L: s.L, B: s.B, setupId: s.id, profileHash: this.hash, confirmations: s.confirmed }
              this.state = 'SIGNALLED'; this.signalledAt = ev
            }
          } else {
            // Retracement or a failed condition resets the confirmation count
            // (the frozen setup stays until it expires).
            s.confirmed = 0; s.dir = null
            if (this.state === 'CONFIRMING') this.state = 'ARMED'
            if (ev - s.armedAt > p.expiryEvents) { this.state = 'EXPIRED'; this.expiredAt = ev }
          }
        }
      }
    } else {
      this.accepted++
    }
    this.mids.push(mid2); this.bids.push(q.bid); this.asks.push(q.ask); this.spreads.push(spread)
    const keep = N + M + 2
    if (this.mids.length > keep) { this.mids.splice(0, this.mids.length - keep); this.bids.splice(0, this.bids.length - keep); this.asks.splice(0, this.asks.length - keep); this.spreads.splice(0, this.spreads.length - keep) }
    return signal
  }
}

/** Runs the oracle over an event list (one symbol) and returns every signal. */
export function runOracle(events, params = {}) {
  const o = new TickMomentumOracle(params)
  const out = []
  for (const q of events) { const s = o.feed(q); if (s) out.push(s) }
  return { signals: out, rejected: o.rejected, accepted: o.accepted, profileHash: o.hash }
}

File: cpp-exec/src/tick_strategy.cpp
// cpp-exec/src/tick_strategy.cpp — see tick_strategy.hpp.
#include "tick_strategy.hpp"

#include <openssl/sha.h>

#include <algorithm>
#include <cmath>
#include <cstdio>

namespace tick {

namespace {
long long roundHalfUp(double x) { return static_cast<long long>(std::floor(x + 0.5)); }
std::string num(double d) {
  // The reference's JSON.stringify for these numbers: integers without a
  // fraction, otherwise the shortest round-trip form (%.17g trimmed).
  if (d == std::floor(d) && std::fabs(d) < 1e15) { char b[32]; std::snprintf(b, sizeof b, "%lld", static_cast<long long>(d)); return b; }
  char b[40];
  for (int prec = 1; prec <= 17; ++prec) {
    std::snprintf(b, sizeof b, "%.*g", prec, d);
    if (std::strtod(b, nullptr) == d) break;
  }
  return b;
}
} // namespace

std::string StrategyParams::canonicalJson() const {
  // Keys in sorted order, as the reference's profileHash builds them.
  std::string s = "{\"id\":\"tick_momentum_breakout\",\"version\":\"v1\"";
  s += ",\"confirmations\":" + num(confirmations);
  s += ",\"expiryEvents\":" + num(expiryEvents);
  s += ",\"maxQuoteAgeMs\":" + num(static_cast<double>(maxQuoteAgeMs));
  s += ",\"maxSpread\":" + num(static_cast<double>(maxSpread));
  s += ",\"minEfficiency\":" + num(minEfficiency);
  s += ",\"minStopPrice\":" + num(static_cast<double>(minStopPrice));
  s += ",\"momentumEvents\":" + num(momentumEvents);
  s += ",\"priceIncrement\":" + num(static_cast<double>(priceIncrement));
  s += ",\"rangeEvents\":" + num(rangeEvents);
  s += ",\"rearmCooldownEvents\":" + num(rearmCooldownEvents);
  s += ",\"spreadBufferMult\":" + num(spreadBufferMult);
  s += ",\"stopVolMult\":" + num(stopVolMult);
  s += "}";
  return s;
}

std::string StrategyParams::profileHash() const {
  const std::string c = canonicalJson();
  unsigned char d[SHA256_DIGEST_LENGTH];
  SHA256(reinterpret_cast<const unsigned char*>(c.data()), c.size(), d);
  char hex[17];
  for (int i = 0; i < 8; ++i) std::snprintf(hex + 2 * i, 3, "%02x", d[i]);
  return std::string(hex, 16);
}

const char* setupStateName(SetupState s) {
  switch (s) {
    case SetupState::WARMING: return "WARMING";
    case SetupState::ARMED: return "ARMED";
    case SetupState::CONFIRMING: return "CONFIRMING";
    case SetupState::SIGNALLED: return "SIGNALLED";
    case SetupState::EXPIRED: return "EXPIRED";
  }
  return "?";
}

TickMomentumStrategy::TickMomentumStrategy(StrategyParams p) : p_(p), hash_(p.profileHash()) {}

void TickMomentumStrategy::invalidate() {
  mids_.clear(); bids_.clear(); asks_.clear(); spreads_.clear();
  diffsN_.clear(); sumSqN_ = 0; lo_.clear(); hi_.clear();
  setup_.reset();
  state_ = SetupState::WARMING;
}

long long TickMomentumStrategy::medianSpread() const {
  const size_t n = lo_.size() + hi_.size();
  if (n == 0) return 0;
  if (n % 2 == 1) return *lo_.rbegin();
  // floor((a + b) / 2) with a <= b, as the reference
  const long long a = *lo_.rbegin(), b = *hi_.begin();
  return static_cast<long long>(std::floor((static_cast<double>(a) + static_cast<double>(b)) / 2.0));
}

void TickMomentumStrategy::pushAccepted(long long mid2, long long bid, long long ask, long long spread) {
  const size_t N = static_cast<size_t>(p_.rangeEvents);
  if (!mids_.empty()) {
    const long long d = mid2 - mids_.back();
    diffsN_.push_back(d);
    sumSqN_ += static_cast<double>(d) * static_cast<double>(d);
    if (diffsN_.size() > N) { const long long old = diffsN_.front(); diffsN_.pop_front(); sumSqN_ -= static_cast<double>(old) * static_cast<double>(old); }
  }
  // spread median over the prior N spreads: insert, evict, rebalance
  if (lo_.empty() || spread <= *lo_.rbegin()) lo_.insert(spread); else hi_.insert(spread);
  spreads_.push_back(spread);
  if (spreads_.size() > N) {
    const long long old = spreads_.front(); spreads_.pop_front();
    auto it = lo_.find(old);
    if (it != lo_.end()) lo_.erase(it); else { auto jt = hi_.find(old); if (jt != hi_.end()) hi_.erase(jt); }
  }
  while (lo_.size() > hi_.size() + 1) { hi_.insert(*lo_.rbegin()); lo_.erase(std::prev(lo_.end())); }
  while (hi_.size() > lo_.size()) { lo_.insert(*hi_.begin()); hi_.erase(hi_.begin()); }
  mids_.push_back(mid2); bids_.push_back(bid); asks_.push_back(ask);
  const size_t keep = N + static_cast<size_t>(p_.momentumEvents) + 2;
  while (mids_.size() > keep) { mids_.pop_front(); bids_.pop_front(); asks_.pop_front(); }
}

std::optional<TickSignal> TickMomentumStrategy::onQuote(const StrategyQuote& q) {
  if (!q.hasBid || !q.hasAsk || q.snapshot || q.crossed) {
    rejected_.invalid++;
    // Plan §4/§5 (11-09-2026 audit): a snapshot — a (re)subscribe, or a
    // continuity break the worker folds into it (gapBefore) — a one-sided
    // update or a crossed quote invalidates the setup AND the warm-up. The
    // window must not carry on with the missed events absent: a fresh N+1
    // prior events are required before anything can arm again.
    invalidate();
    return std::nullopt;
  }
  if (!q.changed) { rejected_.repeat++; return std::nullopt; }
  if (q.ask - q.bid > p_.maxSpread) { rejected_.spread++; return std::nullopt; }
  if (haveLast_ && q.recvMs > lastRecvMs_ && q.recvMs - lastRecvMs_ > static_cast<uint64_t>(p_.maxQuoteAgeMs)) {
    rejected_.stale++;
    invalidate();
    lastRecvMs_ = q.recvMs;
    return std::nullopt;
  }
  haveLast_ = true;
  lastRecvMs_ = q.recvMs;

  const size_t N = static_cast<size_t>(p_.rangeEvents), M = static_cast<size_t>(p_.momentumEvents);
  const long long mid2 = q.bid + q.ask;
  const long long spread = q.ask - q.bid;
  const bool warm = mids_.size() >= N + 1 && mids_.size() >= M;
  std::optional<TickSignal> signal;
  if (warm) {
    // Prior-N range and V (the candidate excluded).
    long long H = mids_[mids_.size() - N], L = H;
    for (size_t i = mids_.size() - N; i < mids_.size(); ++i) { H = std::max(H, mids_[i]); L = std::min(L, mids_[i]); }
    const double V = std::sqrt(static_cast<double>(M) * (sumSqN_ / static_cast<double>(diffsN_.size())));
    const long long medSpread = medianSpread();
    const long long B = std::max<long long>(4 * p_.priceIncrement, roundHalfUp(p_.spreadBufferMult * 2.0 * static_cast<double>(medSpread)));
    // D and E over the last M events including the candidate.
    const long long D = mid2 - mids_[mids_.size() - M];
    long long sumAbs = 0;
    for (size_t i = mids_.size() - M + 1; i < mids_.size(); ++i) sumAbs += std::llabs(mids_[i] - mids_[i - 1]);
    sumAbs += std::llabs(mid2 - mids_.back());
    const double E = sumAbs > 0 ? static_cast<double>(std::llabs(D)) / static_cast<double>(sumAbs) : 0.0;
    accepted_++;
    const uint64_t ev = accepted_;

    if (state_ == SetupState::SIGNALLED) {
      const Setup& s = *setup_;
      if (mid2 >= s.L && mid2 <= s.H && ev - signalledAt_ >= static_cast<uint64_t>(p_.rearmCooldownEvents)) { setup_.reset(); state_ = SetupState::WARMING; }
    } else if (state_ == SetupState::EXPIRED) {
      if (ev - expiredAt_ >= static_cast<uint64_t>(std::max(1, p_.rangeEvents / 4))) { setup_.reset(); state_ = SetupState::WARMING; }
    }

    if (state_ == SetupState::WARMING && !setup_) {
      if (H > L && mid2 >= L && mid2 <= H && std::isfinite(V) && V > 0) {
        Setup s;
        s.H = H; s.L = L; s.B = B; s.armedAt = ev; s.id = ++setupSeq_;
        s.bidHigh = bids_[bids_.size() - N]; s.bidLow = s.bidHigh; s.askHigh = asks_[asks_.size() - N]; s.askLow = s.askHigh;
        for (size_t i = bids_.size() - N; i < bids_.size(); ++i) {
          s.bidHigh = std::max(s.bidHigh, bids_[i]); s.bidLow = std::min(s.bidLow, bids_[i]);
          s.askHigh = std::max(s.askHigh, asks_[i]); s.askLow = std::min(s.askLow, asks_[i]);
        }
        setup_ = s;
        state_ = SetupState::ARMED;
      }
    } else if (state_ == SetupState::ARMED || state_ == SetupState::CONFIRMING) {
      Setup& s = *setup_;
      if (ev - s.armedAt > static_cast<uint64_t>(p_.expiryEvents) && state_ == SetupState::ARMED) {
        state_ = SetupState::EXPIRED; expiredAt_ = ev;
      } else {
        const bool longOk = mid2 > s.H + s.B && D > 0 && E >= p_.minEfficiency && q.bid > s.bidHigh;
        const bool shortOk = mid2 < s.L - s.B && D < 0 && E >= p_.minEfficiency && q.ask < s.askLow;
        const std::string dir = longOk ? "BUY" : (shortOk ? "SELL" : "");
        if (!dir.empty() && (s.dir.empty() || s.dir == dir)) {
          s.dir = dir;
          s.confirmed += 1;
          state_ = SetupState::CONFIRMING;
          if (s.confirmed >= p_.confirmations) {
            TickSignal sig;
            sig.side = dir; sig.seq = q.seq; sig.recvMs = q.recvMs; sig.trigger2 = mid2; sig.bid = q.bid; sig.ask = q.ask;
            sig.stopDistance = std::max<long long>(p_.minStopPrice, roundHalfUp(p_.stopVolMult * V / 2.0));
            sig.spread = spread; sig.V = V; sig.D = D; sig.E = E; sig.H = s.H; sig.L = s.L; sig.B = s.B;
            sig.setupId = s.id; sig.confirmations = s.confirmed;
            signal = sig;
            state_ = SetupState::SIGNALLED; signalledAt_ = ev;
          }
        } else {
          s.confirmed = 0; s.dir.clear();
          if (state_ == SetupState::CONFIRMING) state_ = SetupState::ARMED;
          if (ev - s.armedAt > static_cast<uint64_t>(p_.expiryEvents)) { state_ = SetupState::EXPIRED; expiredAt_ = ev; }
        }
      }
    }
  } else {
    accepted_++;
  }
  pushAccepted(mid2, q.bid, q.ask, spread);
  return signal;
}

} // namespace tick
